layout: Use `Au` in `ResolvedMargins` and `CollapsedMargin`

ektuu opened the pull request:
<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes change ResolvedMargins to use Au instead of length
- [x] These changes do not require tests because  they are a refactoring of internal data structures with no observable effect on functionality.

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

ektuu committed:
change ResolvedMargins to use Au instead of length

mrobinson commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1409,27 +1411,19 @@ fn justify_self_alignment(containing_block: &ContainingBlock, free_space: Au) ->
+        (AuOrAuto::Auto, AuOrAuto::LengthPercentage(end)) => (free_space - end, end),
         (AuOrAuto::LengthPercentage(start), AuOrAuto::Auto) => (start, free_space - start),
         (AuOrAuto::LengthPercentage(start), AuOrAuto::LengthPercentage(end)) => {
-            // In the cases above, the free space is zero after taking 'auto' margins into account.
-            // But here we may still have some free space to perform 'justify-self' alignment.
-            // This aligns the margin box within the containing block, or in other words,
-            // aligns the border box within the margin-shrunken containing block.
```
What is your motivation for removing these comments?

mrobinson commented on components/layout_2020/flow/mod.rs:
```diff
@@ -667,9 +666,9 @@ fn layout_in_flow_non_replaced_block_level_same_formatting_context(
-                "We should know whether we are collapsing the block start margin with the parent \
-                when laying out sequentially",
-            ).0 && style.get_box().clear == Clear::None;
+                 "We should know whether we are collapsing the block start margin with the parent \
+                 when laying out sequentially",
+             ).0 && style.get_box().clear == Clear::None;
```
The white space only changes here indicate that you haven't run './mach fmt' locally before submitting this change. Please do that.

ektuu replied:
I ran ./mach fit multiple times, still this piece of code remain neglected. I've fixed the it manually.

mrobinson replied:
It's pretty unlikely that `./mach fmt` should fail. What is the output you see when running `./mach fmt`? Perhaps there is a problem with the installation of your Rust toolchain?

ektuu replied:
After running ./mach fmt, while most of the codebase's formatting issues are addressed, and indentation changes are correctly applied, there are still some sections of the code that remain unformatted. The command completes without any output messages.

mrobinson replied:
What do you mean by "remain unformatted." The changes that `./mach fmt` does should result in the proper formatting of all Rust code in Servo. Can you please paste the output of running `./mach fmt` and  and then `./mach test-tidy` (which tests that `./mach fmt` was done)? If `./mach test-tidy` fails after running `./mach fmt` then there is definitely a bug that should be fixed. It's quite likely if you modify the formatting of Rust code after running `./mach fmt` though, `./mach test-tidy` will (and should!) fail.

ektuu replied:
<img width="500" alt="Screenshot 2024-03-26 at 00 57 53" src="https://github.com/servo/servo/assets/137225906/1ac1090a-64e4-48ab-9908-cf00d6b6431b">

ektuu replied:
Even while I raised this PR and pushed changes, test-tidy reported no errors.

Loirooriol replied:
Don't change this. Regardless of whether `./mach fmt` addresses this or not.

mrobinson commented on components/layout_2020/flexbox/layout.rs:
```diff
@@ -104,6 +104,10 @@ struct FlexLineLayoutResult {
     item_fragments: Vec<(BoxFragment, PositioningContext)>, // One per flex item, in the given order
 }
 
+fn css_pixel_length_to_au(length: CSSPixelLength) -> Au {
+    Au::from_f32_px(length.px())
+}
```
I'm fairly sure there is already a conversion implementation of Into or From so you shouldn't need to add this helper and can likely just do length_value.into().

mrobinson commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1409,27 +1411,19 @@ fn justify_self_alignment(containing_block: &ContainingBlock, free_space: Au) ->
-        (AuOrAuto::Auto, AuOrAuto::LengthPercentage(end)) => {
-            (Au::zero().max(free_space - end), end)
+            let half = free_space / 2;
+            (half, free_space - half)
```
Why modify the name of the temporary variable here?

mrobinson commented on components/layout_2020/fragment_tree/fragment.rs:
```diff
@@ -279,15 +279,18 @@ impl CollapsedBlockMargins {
 
-    pub fn new(margin: Length) -> Self {
+    pub fn new(margin: Au) -> Self {
+        let max_positive = if margin > Au(0) { margin } else { Au(0) };
+        let min_negative = if margin < Au(0) { margin } else { Au(0) };
```
What's the purpose of changing this code?

mrobinson reviewed (changes requested):
Some of the changes in this PR are a bit strange? Is this code generated with a tool like GitHub Copilot?

ektuu committed:
made suggested changes

removed whitespace changes

made suggested changes

ektuu commented:
@mrobinson I apologise for the mistakes in the code. I was not using GitHub copilot, but I faced some problem with my IDE, therefore there were some unwanted changes. Also there were some formatting changes even after running `./mach fmt` multiple times, I have fixed them manually and also made the other  suggested changes. Please review and provide feedback.
Thanks !

Loirooriol commented on components/layout_2020/flexbox/layout.rs:
```diff
@@ -848,7 +848,15 @@ impl FlexLine<'_> {
                 let content_rect = flex_context.rect_to_flow_relative(line_size, content_rect);
                 let margin = flex_context.sides_to_flow_relative(*margin);
-                let collapsed_margin = CollapsedBlockMargins::from_margin(&margin);
+                let converted_margin = LogicalSides {
+                    block_start: margin.block_start.into(),
+                    block_end: margin.block_end.into(),
+                    inline_start: margin.inline_start.into(),
+                    inline_end: margin.inline_end.into(),
+                };
+
+                let collapsed_margin = CollapsedBlockMargins::from_margin(&converted_margin);
+
```
```suggestion
                let margin = flex_context.sides_to_flow_relative(*margin).into();
                let collapsed_margin = CollapsedBlockMargins::from_margin(&margin);
```

Loirooriol replied:
Please use what I said instead of

```rs
                let margin = flex_context.sides_to_flow_relative(*margin);
                let collapsed_margin =
                    CollapsedBlockMargins::from_margin(&LogicalSides::<Au>::from(margin.clone()));
```

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -242,10 +242,9 @@ impl BlockFormattingContext {
-            content_block_size: (flow_layout.content_block_size +
+            content_block_size: (Au::from(flow_layout.content_block_size) +
                 flow_layout.collapsible_margins_in_children.end.solve() +
-                clearance.unwrap_or_else(Au::zero).into())
-            .into(),
+                clearance.unwrap_or_else(Au::zero)),
```
Remove the surrounding parenthesis.

ektuu replied:
Yeah sure, I will make the changes.

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -765,7 +764,10 @@ fn layout_in_flow_non_replaced_block_level_same_formatting_context(
 
     let block_size = containing_block_for_children.block_size.auto_is(|| {
         content_block_size
-            .clamp_between_extremums(min_box_size.block, max_box_size.block)
+            .clamp_between_extremums(
+                min_box_size.block.into(),
+                max_box_size.block.map(Into::into),
+            )
             .into()
```
Seems simpler as `Length::from(content_block_size).clamp_between_extremums(min_box_size.block,  max_box_size.block)`.

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1369,8 +1371,8 @@ fn solve_margins(
     (
-        pbm.margin.block_start.auto_is(Au::zero),
-        pbm.margin.block_end.auto_is(Au::zero),
+        pbm.margin.block_start.auto_is(|| Au::zero()),
+        pbm.margin.block_end.auto_is(|| Au::zero()),
```
This change is unnecessary.

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -196,7 +196,7 @@ impl BlockLevelBox {
 struct FlowLayout {
     pub fragments: Vec<Fragment>,
-    pub content_block_size: Length,
+    pub content_block_size: Au,
```
This change seems independent from "change ResolvedMargins to use Au instead of length".
I think it would be simpler to change the margins alone.
You could split this into 2 PRs, or otherwise at least mention this in the commit description.

ektuu replied:
I apologise for mixing it with margins. I would stick to changing the margins in this PR.

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1409,27 +1411,22 @@ fn justify_self_alignment(containing_block: &ContainingBlock, free_space: Au) ->
-        containing_block.inline_size - pbm.padding_border_sums.inline - inline_size.into();
-    let mut justification = Au::zero();
+    let free_space = containing_block.inline_size - pbm.padding_border_sums.inline - inline_size;
+    let justification = Au::zero();
```
Do not remove the justification.

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1409,27 +1411,22 @@ fn justify_self_alignment(containing_block: &ContainingBlock, free_space: Au) ->
     let inline_margins = match (pbm.margin.inline_start, pbm.margin.inline_end) {
         (AuOrAuto::Auto, AuOrAuto::Auto) => {
-            let start = Au::zero().max(free_space / 2);
+            let start = free_space / 2;
```
Do not remove the `Au::zero()` minimum, this needs safe centering.

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1409,27 +1411,22 @@ fn justify_self_alignment(containing_block: &ContainingBlock, free_space: Au) ->
-        (AuOrAuto::Auto, AuOrAuto::LengthPercentage(end)) => {
-            (Au::zero().max(free_space - end), end)
-        },
+        (AuOrAuto::Auto, AuOrAuto::LengthPercentage(end)) => (free_space - end, end),
```
Ditto.

Loirooriol replied:
I already said this in a previous review, this needs to do safe alignment, don't change it into unsafe alignment.

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1537,7 +1534,7 @@ impl PlacementState {
             start_margin: CollapsedMargin::zero(),
             current_margin: CollapsedMargin::zero(),
-            current_block_direction_position: Length::zero(),
+            current_block_direction_position: Length::zero().into(),
```
```suggestion
            current_block_direction_position: Au::zero(),
```

Loirooriol commented on components/layout_2020/fragment_tree/fragment.rs:
```diff
@@ -252,14 +252,14 @@ impl IFrameFragment {
         tree.add_item(format!(
             "IFrame\
-                \npipeline={:?} rect={:?}",
+            \npipeline={:?} rect={:?}",
```
Ditto

Loirooriol replied:
You are still changing the indentation.

ektuu committed:
Update components/layout_2020/flexbox/layout.rs

Co-authored-by: Oriol Brufau <obrufau@igalia.com>

ektuu committed:
Update components/layout_2020/flow/mod.rs

Co-authored-by: Oriol Brufau <obrufau@igalia.com>

ektuu commented on components/layout_2020/fragment_tree/fragment.rs:
```diff
@@ -279,15 +279,15 @@ impl CollapsedBlockMargins {
         Self {
-            max_positive: Length::zero(),
-            min_negative: Length::zero(),
+            max_positive: Au(0),
+            min_negative: Au(0),
```
When using Au::zero(), I received an error indicating that the method does not exist for Au.
After switching to num_traits::Zero, I faced an unresolved import error for num_traits, despite confirming the crate's inclusion in Cargo.toml and double-checking for typos.Could you please provide any insights into what might be causing these issues or suggest alternative approaches?

Loirooriol replied:
It doesn't work because you are removing `use style::Zero;` at the top of this file.

Loirooriol commented on components/layout_2020/flexbox/layout.rs:
```diff
@@ -848,6 +848,7 @@ impl FlexLine<'_> {
             .map(|(((item, item_result), content_rect), margin)| {
                 let content_rect = flex_context.rect_to_flow_relative(line_size, content_rect);
                 let margin = flex_context.sides_to_flow_relative(*margin);
+                let margin = flex_context.sides_to_flow_relative(*margin).into();
```
Remove the previous line, this doesn't compile.

Also, the `margin.into()` below should now be just `margin`.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

ektuu committed:
fixed errors

fixed errors

fixed formatting

fixed errors

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

mrobinson commented on components/layout_2020/flow/mod.rs:
```diff
@@ -764,7 +763,8 @@ fn layout_in_flow_non_replaced_block_level_same_formatting_context(
 
     let block_size = containing_block_for_children.block_size.auto_is(|| {
-        content_block_size
+        content_block_size;
```
It doesn't look like this line is doing anything. It's just a variable followed by a semicolon.

mrobinson commented on components/layout_2020/flow/mod.rs:
```diff
@@ -242,10 +242,9 @@ impl BlockFormattingContext {
         IndependentLayout {
             fragments: flow_layout.fragments,
-            content_block_size: (flow_layout.content_block_size +
+            content_block_size: Au::from(flow_layout.content_block_size) +
```
Can this just be `flow_layout.content_block_size.into()`?

ektuu replied:
Due to compiler errors with .into(), which could not infer the correct type conversion, I've used Au::from(...) for explicit and clear type conversion.

mrobinson commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1537,7 +1534,7 @@ impl PlacementState {
             start_margin: CollapsedMargin::zero(),
             current_margin: CollapsedMargin::zero(),
-            current_block_direction_position: Length::zero(),
+            current_block_direction_position: Au::zero().into(),
```
```suggestion
            current_block_direction_position: Length::zero(),
```

mrobinson commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1609,7 +1606,7 @@ impl PlacementState {
 
                 if self.next_in_flow_margin_collapses_with_parent_start_margin {
-                    debug_assert_eq!(self.current_margin.solve(), Length::zero());
+                    debug_assert_eq!(self.current_margin.solve(), Au::zero());
```
```suggestion
                    debug_assert!(self.current_margin.solve().is_zero());
```

mrobinson commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1621,9 +1618,10 @@ impl PlacementState {
                         .adjoin_assign(&fragment_block_margins.start);
                 }
+                self.current_block_direction_position += self.current_margin.solve().into();
 
                 fragment.content_rect.start_corner.block +=
-                    self.current_margin.solve() + self.current_block_direction_position;
+                    self.current_block_direction_position.into();
```
Please keep these together.

```
fragment.content_rect.start_corner.block += (self.current_margin.solve() + self.current_block_direction_position).into();

mrobinson commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1632,23 +1630,23 @@ impl PlacementState {
                 } else {
-                    self.current_block_direction_position +=
-                        self.current_margin.solve() + fragment_block_size.into();
+                    self.current_block_direction_position += fragment_block_size.into();
```
You've modified the code here. You are no longer adding `self.current_margin.solve()`.

mrobinson commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1632,23 +1630,23 @@ impl PlacementState {
                     .expect("Found float fragment without SequentialLayoutState");
-                let block_offset_from_containing_block_top =
-                    self.current_block_direction_position + self.current_margin.solve();
+                self.current_block_direction_position += self.current_margin.solve().into();
+                let block_offset_from_containing_block_top = self.current_block_direction_position;
```
This also looks quite different from the original code. Before this code was not modifying `self.current_block_direction_position`, but now it is. Why make this change?

ektuu replied:
I apologise. I have addressed this and reverted unnecessary changes.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

Loirooriol commented:
I would rebase and squash your commits. https://github.com/servo/servo/pull/31848/commits has 28 commits, most are empty or unrelated, and the empty ones are causing these WPT export errors. In general, you shouldn't click  the "Update branch button".

ektuu committed:
modified changes

made suggested changes

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

Loirooriol commented on components/layout_2020/flow/inline.rs:
```diff
@@ -1678,7 +1678,7 @@ impl InlineFormattingContext {
         FlowLayout {
             fragments: ifc.fragments,
-            content_block_size,
+            content_block_size: content_block_size.into(),
```
Don't change this.

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -441,7 +440,7 @@ fn layout_block_level_children(
     FlowLayout {
         fragments,
-        content_block_size,
+        content_block_size: content_block_size.into(),
```
Don't change this.

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -764,7 +763,7 @@ fn layout_in_flow_non_replaced_block_level_same_formatting_context(
 
     let block_size = containing_block_for_children.block_size.auto_is(|| {
-        content_block_size
+        Length::from(content_block_size)
```
Don't change this.

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1428,8 +1426,6 @@ fn solve_inline_margins_for_in_flow_block_level(
             // This aligns the margin box within the containing block, or in other words,
             // aligns the border box within the margin-shrunken containing block.
-            let free_space = Au::zero().max(free_space - start - end);
-            justification = justify_self_alignment(containing_block, free_space);
```
Don't remove this.

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1609,7 +1605,7 @@ impl PlacementState {
 
                 if self.next_in_flow_margin_collapses_with_parent_start_margin {
-                    debug_assert_eq!(self.current_margin.solve(), Length::zero());
+                    debug_assert_eq!(self.current_margin.solve().is_zero(), true);
```
If the assert fails, I think `debug_assert_eq!(self.current_margin.solve(), Au::zero());` will provide the non-zero value, so it could be more useful.

Or otherwise, simplify to `debug_assert!(self.current_margin.solve().is_zero());`

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1621,9 +1617,11 @@ impl PlacementState {
                     self.current_margin
                         .adjoin_assign(&fragment_block_margins.start);
                 }
+                self.current_block_direction_position += self.current_margin.solve().into();
```
Don't add this.

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1621,9 +1617,11 @@ impl PlacementState {
 
-                fragment.content_rect.start_corner.block +=
-                    self.current_margin.solve() + self.current_block_direction_position;
+                fragment.content_rect.start_corner.block += (self.current_margin.solve() +
+                    self.current_block_direction_position.into())
+                .into();
```
Avoid back-and-forth conversions:

```suggestion
                fragment.content_rect.start_corner.block +=
                    Length::from(self.current_margin.solve()) +
                        self.current_block_direction_position;
```

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1633,22 +1631,23 @@ impl PlacementState {
                 } else {
                     self.current_block_direction_position +=
-                        self.current_margin.solve() + fragment_block_size.into();
+                        (self.current_margin.solve() + fragment_block_size.into()).into();
```
```suggestion
                        (self.current_margin.solve() + fragment_block_size).into();
```

Loirooriol replied:
You didn't change this.

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1662,11 +1661,11 @@ impl PlacementState {
         }
         (
-            self.current_block_direction_position,
+            self.current_block_direction_position.into(),
```
Don't change this.

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1633,22 +1631,23 @@ impl PlacementState {
                 let offset = LogicalVec2 {
-                    block: self.current_margin.solve() + self.current_block_direction_position,
-                    inline: Length::new(0.),
+                    block: self.current_margin.solve() +
+                        self.current_block_direction_position.into(),
+                    inline: Au::from_px(0),
                 };
-                fragment.borrow_mut().adjust_offsets(offset);
+                fragment.borrow_mut().adjust_offsets(offset.into());
```
This seems a smaller diff:

```rs
                let offset = LogicalVec2 {
                    block: Length::from(self.current_margin.solve()) +
                        self.current_block_direction_position,
                    inline: Length::new(0.),
                };
                fragment.borrow_mut().adjust_offsets(offset);
```

Loirooriol commented on components/layout_2020/fragment_tree/fragment.rs:
```diff
@@ -279,15 +280,15 @@ impl CollapsedBlockMargins {
         Self {
-            max_positive: margin.max(Length::zero()),
-            min_negative: margin.min(Length::zero()),
+            max_positive: std::cmp::max(margin, Au::zero()),
+            min_negative: std::cmp::min(margin, Au::zero()),
```
No need to switch to `std::cmp::max`.

```suggestion
            max_positive: margin.max(Au::zero()),
            min_negative: margin.max(Au::zero()),
```

Loirooriol commented on components/layout_2020/fragment_tree/hoisted_shared_fragment.rs:
```diff
@@ -61,7 +61,7 @@ impl AbsoluteBoxOffsets {
     pub(crate) fn adjust_offset(&mut self, new_offset: Length) {
         if let AbsoluteBoxOffsets::StaticStart { ref mut start } = *self {
-            *start = new_offset
+            *start = new_offset.into();
```
Don't change this.

ektuu committed:
Update components/layout_2020/fragment_tree/fragment.rs

Co-authored-by: Oriol Brufau <obrufau@igalia.com>

updated components/layout_2020/fragment_tree/fragment.rs

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

ektuu committed:
Update components/layout_2020/flow/mod.rs

Co-authored-by: Oriol Brufau <obrufau@igalia.com>

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

ektuu committed:
Update components/layout_2020/flow/mod.rs

Co-authored-by: Oriol Brufau <obrufau@igalia.com>

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

ektuu committed:
updated changes

updated changes

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

ektuu committed:
unified all the suggested changes here

ektuu committed:
fixed formatting

eerii commented on components/layout_2020/flow/mod.rs:
```diff
@@ -242,10 +242,9 @@ impl BlockFormattingContext {
                 flow_layout.collapsible_margins_in_children.end.solve() +
-                clearance.unwrap_or_else(Au::zero).into())
-            .into(),
+                clearance.unwrap_or_else(Au::zero),
```
I think clippy might complain about using or else when `Au` has a [`Default`](https://docs.rs/app_units/latest/app_units/struct.Au.html#impl-Default-for-Au) implementation that sets it to zero. You could do something like:
```suggestion
                clearance.unwrap_or_default(),
```

eerii commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1632,14 +1629,14 @@ impl PlacementState {
                 } else {
-                    self.current_block_direction_position +=
-                        self.current_margin.solve() + fragment_block_size.into();
+                    (self.current_margin.solve() + fragment_block_size.into());
```
Is this addition saved anywhere? The `self.current_block_direction_position +=` from before seems to be missing, you may need to revert this change.

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -670,6 +669,7 @@ fn layout_in_flow_non_replaced_block_level_same_formatting_context(
                 "We should know whether we are collapsing the block start margin with the parent \
                 when laying out sequentially",
             ).0 && style.get_box().clear == Clear::None;
+
```
Don't add a newline

Loirooriol commented on components/layout_2020/flow/mod.rs:
```diff
@@ -1409,27 +1409,24 @@ fn justify_self_alignment(containing_block: &ContainingBlock, free_space: Au) ->
             // aligns the border box within the margin-shrunken containing block.
             let free_space = Au::zero().max(free_space - start - end);
-            justification = justify_self_alignment(containing_block, free_space);
+            let mut justification = justify_self_alignment(containing_block, free_space);
```
Do not change this. This must be setting the existing variable, not create a local one.

Loirooriol commented on components/layout_2020/fragment_tree/fragment.rs:
```diff
@@ -279,15 +280,15 @@ impl CollapsedBlockMargins {
-            max_positive: margin.max(Length::zero()),
-            min_negative: margin.min(Length::zero()),
+            max_positive: margin.max(Au::zero()),
+            min_negative: margin.max(Au::zero()),
```
Do not change the min into a max.

Loirooriol commented:
Your PR keeps having accidental removals, accidental additions, or accidental changes.
Before pushing, please run `git diff main` and check that the changes make sense.

ektuu commented:
> Your PR keeps having accidental removals, accidental additions, or accidental changes. Before pushing, please run `git diff main` and check that the changes make sense.

I apologize for the accidental changes in the recent PRs due to IDE and GitHub issues, which are now resolved. I assure you of more careful reviews moving forward to prevent recurrence.

ektuu committed:
reverted unnecessary changes

Loirooriol commented on components/layout_2020/fragment_tree/fragment.rs:
```diff
@@ -242,7 +242,7 @@ impl ImageFragment {
         tree.add_item(format!(
             "Image\
-                \nrect={:?}",
+            \nrect={:?}",
```
You are still changing the indentation.

Loirooriol reviewed (changes requested):
Compare `./mach cargo-clippy -p layout_2020` on main vs on your patch. It's introducing some warnings for type conversions that are no longer necessary.

```
warning: useless conversion to the same type: `geom::LogicalSides<app_units::Au>`
   --> components/layout_2020/flexbox/layout.rs:861:25
    |
861 |                         margin.into(),
    |                         ^^^^^^^^^^^^^ help: consider removing `.into()`: `margin`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
    = note: `#[warn(clippy::useless_conversion)]` on by default

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/float.rs:1065:44
     |
1065 |         self.bfc_relative_block_position + self.current_margin.solve().into()
     |                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `self.current_margin.solve()`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/float.rs:1074:37
     |
1074 |         self.advance_block_position(self.current_margin.solve().into());
     |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `self.current_margin.solve()`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/float.rs:1083:13
     |
1083 | /             self.current_margin
1084 | |                 .adjoin(block_start_margin)
1085 | |                 .solve()
1086 | |                 .into()
     | |_______________________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
help: consider removing `.into()`
     |
1083 ~             self.current_margin
1084 +                 .adjoin(block_start_margin)
1085 +                 .solve()
     |

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/float.rs:1095:13
     |
1095 |             self.current_margin.solve().into() +
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `self.current_margin.solve()`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/float.rs:1096:13
     |
1096 |             block_start_margin.solve().into()
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `block_start_margin.solve()`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/float.rs:1190:13
     |
1190 | /             self.floats
1191 | |                 .containing_block_info
1192 | |                 .block_start_margins_not_collapsed
1193 | |                 .solve()
1194 | |                 .into()
     | |_______________________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
help: consider removing `.into()`
     |
1190 ~             self.floats
1191 +                 .containing_block_info
1192 +                 .block_start_margins_not_collapsed
1193 +                 .solve()
     |

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/float.rs:1205:13
     |
1205 | /             self.floats
1206 | |                 .containing_block_info
1207 | |                 .block_start_margins_not_collapsed
1208 | |                 .adjoin(&margins_collapsing_with_parent_containing_block)
1209 | |                 .solve()
1210 | |                 .into();
     | |_______________________^
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
help: consider removing `.into()`
     |
1205 ~             self.floats
1206 +                 .containing_block_info
1207 +                 .block_start_margins_not_collapsed
1208 +                 .adjoin(&margins_collapsing_with_parent_containing_block)
1209 ~                 .solve();
     |

warning: useless conversion to the same type: `app_units::Au`
   --> components/layout_2020/flow/mod.rs:120:62
    |
120 |         collected_margin.adjoin_assign(&CollapsedMargin::new(start_margin.into()));
    |                                                              ^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `start_margin`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
   --> components/layout_2020/flow/mod.rs:179:62
    |
179 |         collected_margin.adjoin_assign(&CollapsedMargin::new(end_margin.into()));
    |                                                              ^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `end_margin`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
   --> components/layout_2020/flow/mod.rs:640:9
    |
640 |         containing_block_for_children.inline_size.into(),
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `containing_block_for_children.inline_size`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
   --> components/layout_2020/flow/mod.rs:712:17
    |
712 |                 effective_margin_inline_start.into();
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `effective_margin_inline_start`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
   --> components/layout_2020/flow/mod.rs:803:17
    |
803 |                 effective_margin_inline_start.into(),
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `effective_margin_inline_start`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `geom::LogicalSides<app_units::Au>`
   --> components/layout_2020/flow/mod.rs:818:9
    |
818 |         margin.into(),
    |         ^^^^^^^^^^^^^ help: consider removing `.into()`: `margin`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
   --> components/layout_2020/flow/mod.rs:880:51
    |
880 |         } = solve_margins(containing_block, &pbm, inline_size.into());
    |                                                   ^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `inline_size`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
   --> components/layout_2020/flow/mod.rs:887:21
    |
887 |                     effective_margin_inline_start.into(),
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `effective_margin_inline_start`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `geom::LogicalSides<app_units::Au>`
   --> components/layout_2020/flow/mod.rs:904:13
    |
904 |             margin.into(),
    |             ^^^^^^^^^^^^^ help: consider removing `.into()`: `margin`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
   --> components/layout_2020/flow/mod.rs:937:65
    |
937 |         let collapsed_margin_block_start = CollapsedMargin::new(margin_block_start.into());
    |                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_block_start`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/mod.rs:1114:27
     |
1114 |             inline_start: margin_inline_start.into(),
     |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_inline_start`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/mod.rs:1115:25
     |
1115 |             inline_end: margin_inline_end.into(),
     |                         ^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_inline_end`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/mod.rs:1116:26
     |
1116 |             block_start: margin_block_start.into(),
     |                          ^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_block_start`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/mod.rs:1117:24
     |
1117 |             block_end: margin_block_end.into(),
     |                        ^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_block_end`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `geom::LogicalSides<app_units::Au>`
    --> components/layout_2020/flow/mod.rs:1156:13
     |
1156 |             margin.into(),
     |             ^^^^^^^^^^^^^ help: consider removing `.into()`: `margin`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/mod.rs:1195:65
     |
1195 |         let collapsed_margin_block_start = CollapsedMargin::new(margin_block_start.into());
     |                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_block_start`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/mod.rs:1221:69
     |
1221 |         sequential_layout_state.adjoin_assign(&CollapsedMargin::new(margin_block_end.into()));
     |                                                                     ^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_block_end`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/mod.rs:1230:13
     |
1230 |             content_size.inline.into(),
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `content_size.inline`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/mod.rs:1235:23
     |
1235 |         inline_start: margin_inline_start.into(),
     |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_inline_start`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/mod.rs:1236:21
     |
1236 |         inline_end: margin_inline_end.into(),
     |                     ^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_inline_end`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/mod.rs:1237:22
     |
1237 |         block_start: margin_block_start.into(),
     |                      ^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_block_start`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `app_units::Au`
    --> components/layout_2020/flow/mod.rs:1238:20
     |
1238 |         block_end: margin_block_end.into(),
     |                    ^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_block_end`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

warning: useless conversion to the same type: `geom::LogicalSides<app_units::Au>`
    --> components/layout_2020/flow/mod.rs:1261:9
     |
1261 |         margin.into(),
     |         ^^^^^^^^^^^^^ help: consider removing `.into()`: `margin`
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
```

ektuu committed:
fixed unnecessary warnings

ektuu commented:
> Compare `./mach cargo-clippy -p layout_2020` on main vs on your patch. It's introducing some warnings for type conversions that are no longer necessary.
> 
> ```
> warning: useless conversion to the same type: `geom::LogicalSides<app_units::Au>`
>    --> components/layout_2020/flexbox/layout.rs:861:25
>     |
> 861 |                         margin.into(),
>     |                         ^^^^^^^^^^^^^ help: consider removing `.into()`: `margin`
>     |
>     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
>     = note: `#[warn(clippy::useless_conversion)]` on by default
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/float.rs:1065:44
>      |
> 1065 |         self.bfc_relative_block_position + self.current_margin.solve().into()
>      |                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `self.current_margin.solve()`
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/float.rs:1074:37
>      |
> 1074 |         self.advance_block_position(self.current_margin.solve().into());
>      |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `self.current_margin.solve()`
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/float.rs:1083:13
>      |
> 1083 | /             self.current_margin
> 1084 | |                 .adjoin(block_start_margin)
> 1085 | |                 .solve()
> 1086 | |                 .into()
>      | |_______________________^
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> help: consider removing `.into()`
>      |
> 1083 ~             self.current_margin
> 1084 +                 .adjoin(block_start_margin)
> 1085 +                 .solve()
>      |
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/float.rs:1095:13
>      |
> 1095 |             self.current_margin.solve().into() +
>      |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `self.current_margin.solve()`
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/float.rs:1096:13
>      |
> 1096 |             block_start_margin.solve().into()
>      |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `block_start_margin.solve()`
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/float.rs:1190:13
>      |
> 1190 | /             self.floats
> 1191 | |                 .containing_block_info
> 1192 | |                 .block_start_margins_not_collapsed
> 1193 | |                 .solve()
> 1194 | |                 .into()
>      | |_______________________^
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> help: consider removing `.into()`
>      |
> 1190 ~             self.floats
> 1191 +                 .containing_block_info
> 1192 +                 .block_start_margins_not_collapsed
> 1193 +                 .solve()
>      |
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/float.rs:1205:13
>      |
> 1205 | /             self.floats
> 1206 | |                 .containing_block_info
> 1207 | |                 .block_start_margins_not_collapsed
> 1208 | |                 .adjoin(&margins_collapsing_with_parent_containing_block)
> 1209 | |                 .solve()
> 1210 | |                 .into();
>      | |_______________________^
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> help: consider removing `.into()`
>      |
> 1205 ~             self.floats
> 1206 +                 .containing_block_info
> 1207 +                 .block_start_margins_not_collapsed
> 1208 +                 .adjoin(&margins_collapsing_with_parent_containing_block)
> 1209 ~                 .solve();
>      |
> 
> warning: useless conversion to the same type: `app_units::Au`
>    --> components/layout_2020/flow/mod.rs:120:62
>     |
> 120 |         collected_margin.adjoin_assign(&CollapsedMargin::new(start_margin.into()));
>     |                                                              ^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `start_margin`
>     |
>     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>    --> components/layout_2020/flow/mod.rs:179:62
>     |
> 179 |         collected_margin.adjoin_assign(&CollapsedMargin::new(end_margin.into()));
>     |                                                              ^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `end_margin`
>     |
>     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>    --> components/layout_2020/flow/mod.rs:640:9
>     |
> 640 |         containing_block_for_children.inline_size.into(),
>     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `containing_block_for_children.inline_size`
>     |
>     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>    --> components/layout_2020/flow/mod.rs:712:17
>     |
> 712 |                 effective_margin_inline_start.into();
>     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `effective_margin_inline_start`
>     |
>     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>    --> components/layout_2020/flow/mod.rs:803:17
>     |
> 803 |                 effective_margin_inline_start.into(),
>     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `effective_margin_inline_start`
>     |
>     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `geom::LogicalSides<app_units::Au>`
>    --> components/layout_2020/flow/mod.rs:818:9
>     |
> 818 |         margin.into(),
>     |         ^^^^^^^^^^^^^ help: consider removing `.into()`: `margin`
>     |
>     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>    --> components/layout_2020/flow/mod.rs:880:51
>     |
> 880 |         } = solve_margins(containing_block, &pbm, inline_size.into());
>     |                                                   ^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `inline_size`
>     |
>     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>    --> components/layout_2020/flow/mod.rs:887:21
>     |
> 887 |                     effective_margin_inline_start.into(),
>     |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `effective_margin_inline_start`
>     |
>     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `geom::LogicalSides<app_units::Au>`
>    --> components/layout_2020/flow/mod.rs:904:13
>     |
> 904 |             margin.into(),
>     |             ^^^^^^^^^^^^^ help: consider removing `.into()`: `margin`
>     |
>     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>    --> components/layout_2020/flow/mod.rs:937:65
>     |
> 937 |         let collapsed_margin_block_start = CollapsedMargin::new(margin_block_start.into());
>     |                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_block_start`
>     |
>     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/mod.rs:1114:27
>      |
> 1114 |             inline_start: margin_inline_start.into(),
>      |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_inline_start`
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/mod.rs:1115:25
>      |
> 1115 |             inline_end: margin_inline_end.into(),
>      |                         ^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_inline_end`
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/mod.rs:1116:26
>      |
> 1116 |             block_start: margin_block_start.into(),
>      |                          ^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_block_start`
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/mod.rs:1117:24
>      |
> 1117 |             block_end: margin_block_end.into(),
>      |                        ^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_block_end`
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `geom::LogicalSides<app_units::Au>`
>     --> components/layout_2020/flow/mod.rs:1156:13
>      |
> 1156 |             margin.into(),
>      |             ^^^^^^^^^^^^^ help: consider removing `.into()`: `margin`
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/mod.rs:1195:65
>      |
> 1195 |         let collapsed_margin_block_start = CollapsedMargin::new(margin_block_start.into());
>      |                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_block_start`
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/mod.rs:1221:69
>      |
> 1221 |         sequential_layout_state.adjoin_assign(&CollapsedMargin::new(margin_block_end.into()));
>      |                                                                     ^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_block_end`
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/mod.rs:1230:13
>      |
> 1230 |             content_size.inline.into(),
>      |             ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `content_size.inline`
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/mod.rs:1235:23
>      |
> 1235 |         inline_start: margin_inline_start.into(),
>      |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_inline_start`
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/mod.rs:1236:21
>      |
> 1236 |         inline_end: margin_inline_end.into(),
>      |                     ^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_inline_end`
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/mod.rs:1237:22
>      |
> 1237 |         block_start: margin_block_start.into(),
>      |                      ^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_block_start`
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `app_units::Au`
>     --> components/layout_2020/flow/mod.rs:1238:20
>      |
> 1238 |         block_end: margin_block_end.into(),
>      |                    ^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `margin_block_end`
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> 
> warning: useless conversion to the same type: `geom::LogicalSides<app_units::Au>`
>     --> components/layout_2020/flow/mod.rs:1261:9
>      |
> 1261 |         margin.into(),
>      |         ^^^^^^^^^^^^^ help: consider removing `.into()`: `margin`
>      |
>      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
> ```

I've addressed all the unnecessary warnings and fixed them.

ektuu commented:
@mrobinson @Loirooriol 
Thank you for guiding me through this update. I learned a lot and apologize for any missteps along the way. Could you please assign me another issue to apply what I've learned?
